File: ODESolver/eParse.c
#include "eParse.h"

char *eq_buffer;
int len_buffer;
int curr_index;
int eq_degree;

char* CONS[LEN_CONS] = {"e\0","pi\0"};
char* VAR[LEN_VAR] = {"u0\0","t\0"};
char* DERIV[LEN_DERIV] = {"u0\0","u1\0","u2\0","u3\0","u4\0","u5\0"};
char* OP[LEN_OP] = {"+\0","-\0","*\0","/\0","^\0"};

double cons_def[2] = {M_E,M_PI};
double var_def[2] = {0.0,0.0};
double deriv_def[6] = {0.0,0.0,0.0,0.0,0.0,0.0};

void print_spaces(int depth){
	int i;
	for(i = 0 ;i < depth;i++)
		printf(" ");
}

char* extract_string(char* str,int init,int final){
	char* tmp;
	int idx;
	tmp = (char*)malloc(sizeof(char)*MAX_STR_LENGTH);
	for(idx=0;idx<(final - init);idx++)
		tmp[idx] = str[init + idx];
	tmp[idx] = '\0';
	return tmp;
}

char* remove_spaces(char* s){
	char* tmp = (char*) malloc(sizeof(char)*MAX_EQ_LENGTH);
	int idx = 0;
	while(*s != '\0' && *s!='\n'){
		if(*s != ' '){
			tmp[idx] = *s;
			idx++;
		}
		s++;
	}
	tmp[idx] = '\0';
	return tmp;
}

char* replace_char(char* s, char c1, char c2){
	char* tmp = (char*) malloc(sizeof(char)*MAX_EQ_LENGTH);
	int idx=0;
	while(*s != '\0' && *s!='\n'){
		if(*s == c1)
			tmp[idx] = c2;
		else
			tmp[idx] = *s;
		s++;\
		idx++;
	}
	tmp[idx] = '\0';
	return tmp;
}

int is_number(char c){
	return c >= '0' && c <= '9' || c=='.';
}

int is_letter(char c){
	return c >= 'a' && c <= 'z';
}

int len_str(char* s){
	int i=0;
	while(s[i]!='\0' && s[i]!='\n')
		i++;
	return i;
}

int cmp_str(char* str1, char* str2){
	while(*str1 != '\0' && *str2 != '\0'){
		if(*str1 != *str2)
			return 0;
		str1++;
		str2++;
	}
	if(*str1 == '\0' && *str2 == '\0')
		return 1;
	return 0;
}

int is_constant(char* s){
	int i;
	for(i=0;i<LEN_CONS;i++){
		if(cmp_str(s,CONS[i]))
			return 1;
	}
	return 0;
}

int is_variable(char c){
	if(c =='x')
		return 1;
	else if(c == 't')
		return 2;
	return 0;
}

int char2number(char c){
	return c - '0';
}

int is_operation(char c){
	switch(c){
		case '+':
			return 1;
		case '-':
			return 2;
		case '*':
			return 3;
		case '/':
			return 4;
		case '^':
			return 5;	
	}
	return 0;
}

int is_term(char c){
	return c=='d' || c=='f' || c=='c' || is_number(c) || c=='(' || is_variable(c);
}

char** split_string(char* s, char delim){
	int i=0, n=0, prev=0;
	char** out = (char**) malloc(sizeof(char*));
	while(s[i]!='\0' && s[i]!='\n'){
		if(s[i] == delim){
			out[n] = extract_string(s,prev,i);
			n++;
			out = (char**) realloc(out,sizeof(char*)*n);
			prev = i+1;
		}
		i++;
	}
	out[n] = extract_string(s,prev,i);
	return out;
}

void set_variable(int idx, double value){
	var_def[idx] = value;
}

void set_derivative(int idx, double value){
	deriv_def[idx] = value;
}

double get_variable(int idx){
	return var_def[idx];
}

double get_derivative(int idx){
	return deriv_def[idx];
}

int get_eq_degree(){
	return eq_degree;
}

enode* create_enode(char* type){
	enode *tmp;
	tmp = (enode*) malloc(sizeof(enode));
	tmp->type = type;
	tmp->child = NULL;
	tmp->num_child = 0;
	return tmp;
}

void add_child(enode* root, enode* child){
	if(root->child == NULL)
		root->child = (enode**) malloc(sizeof(enode*));
	root->child = (enode**)realloc(root->child,sizeof(enode*)*(root->num_child+1));
	root->child[root->num_child] = child;
	root->num_child += 1;
}
enode* pop_child(enode* root){
	root->num_child--;
	return root->child[root->num_child];
}

void print_tree(enode* root, int depth){
	int i;
	print_spaces(depth);
	printf("%s\n",root->type);
	for(i=0;i<root->num_child;i++)
		print_tree(root->child[i],depth+1);
}	

void init_parsing(char* eq){
	int i;
	eq_buffer = remove_spaces(eq);
	curr_index = 0;
	len_buffer = len_str(eq_buffer);
	eq_degree = 0;
}

void print_error(char* s,int i){
	printf("Parsing Error when parsing %s\n at index %d\n",s,i);
	exit(0);
}

enode* parse_constant(){ // const_(name)
	char *tmp;
	int idx;
	enode* out = create_enode("constant");
	enode* aux;
	curr_index += 6;
	idx = curr_index;
	while(is_letter(eq_buffer[curr_index])){
		curr_index++;
		if(curr_index==len_buffer)
			break;
	}
	tmp = extract_string(eq_buffer,idx,curr_index);
	if(is_constant(tmp)){
		aux = create_enode(tmp);
		add_child(out,aux);
	}
	else
		print_error("constant",curr_index);
	return out;
}

enode* parse_variable(){ // x or t
	int tmp = is_variable(eq_buffer[curr_index]);
	enode* out = create_enode("variable");
	enode* aux;
	curr_index++;
	if(tmp){
		aux = create_enode(VAR[tmp-1]);
		add_child(out,aux);
	}
	else
		print_error("variable",curr_index);
	return out;
}

enode* parse_number(){
	enode* out = create_enode("number");
	enode* aux;
	char* tmp;
	int idx;
	idx = curr_index;
	while(is_number(eq_buffer[curr_index])){
		curr_index++;
		if(curr_index == len_buffer)
			break;
	}
	tmp = extract_string(eq_buffer,idx,curr_index);
	aux = create_enode(tmp);
	add_child(out,aux);
	return out;
}

enode* parse_function(){ // func_(name)
	enode* out = create_enode("function");
	enode* aux;
	enode* aux1;
	char* tmp;
	int idx;
	curr_index += 5;
	idx = curr_index;
	while(is_letter(eq_buffer[curr_index])){
		curr_index++;
		if(curr_index == len_buffer)
			break;
	}
	tmp = extract_string(eq_buffer,idx,curr_index);
	aux = create_enode(tmp);
	if(eq_buffer[curr_index]!='(')
		print_error("function",curr_index);
	curr_index++;
	aux1 = parse_expression();
	if(eq_buffer[curr_index]!=')')
		print_error("function",curr_index);
	curr_index++;
	add_child(aux,aux1);
	add_child(out,aux);
	return out;
}

enode* parse_derivative(){ // dkx/dtk | dx/dt
	enode* out = create_enode("derivative");
	enode* aux;
	int degree;
	curr_index++;
	if(is_number(eq_buffer[curr_index])){
		degree = char2number(eq_buffer[curr_index]);
		curr_index += 6;
	}
	else{
		degree = 1;
		curr_index += 4;
	}
	if(degree>eq_degree)
		eq_degree = degree;
	aux = create_enode(DERIV[degree]);
	add_child(out,aux);
	return out;
}

enode* parse_term(){
	enode* out = create_enode("term");
	enode* aux;
	enode* tmp;
	enode* aux1;
	while(eq_buffer[curr_index] != '+' && eq_buffer[curr_index] != '-' && eq_buffer[curr_index] != ')'){
		if(is_number(eq_buffer[curr_index])){
			aux = parse_number();
			add_child(out,aux);
		}
		else if(eq_buffer[curr_index] == 'c'){
			aux = parse_constant();
			add_child(out,aux);
		}
		else if(is_variable(eq_buffer[curr_index])){
			aux = parse_variable();
			add_child(out,aux);
		}
		else if(eq_buffer[curr_index] == 'f'){
			aux = parse_function();
			add_child(out,aux);
		}
		else if(eq_buffer[curr_index] == 'd'){
			aux = parse_derivative();
			add_child(out,aux);
		}
		else if(eq_buffer[curr_index] == '('){
			curr_index++;
			aux = parse_expression();
			add_child(out,aux);
			if(eq_buffer[curr_index] != ')')
				print_error("term",curr_index);
			curr_index++;
		}
		else if(is_operation(eq_buffer[curr_index]) > 2){
			int opt = is_operation(eq_buffer[curr_index]);
			aux = create_enode("operation");
			tmp = create_enode(OP[opt-1]);
			if(out->num_child > 0)
				add_child(tmp,pop_child(out));
			else
				print_error("operation",curr_index);
			curr_index++;
			aux1 = parse_term();
			add_child(tmp,aux1);
			add_child(aux,tmp);
			add_child(out,aux);
		}
		if(curr_index == len_buffer)
			break;
	}
	return out;
}

enode* parse_expression(){
	enode* out = create_enode("expression");
	enode* tmp;
	enode* aux;
	enode* aux1;
	enode* aux2;
	while(curr_index < len_buffer){
		if(is_operation(eq_buffer[curr_index])){
			tmp = create_enode("operation");
			aux = create_enode(OP[is_operation(eq_buffer[curr_index]) - 1]);
			if(out->num_child == 0){
				if(eq_buffer[curr_index] == '-'){
					aux2 = create_enode("term");
					aux1 = create_enode("number");
					add_child(aux1,create_enode("0"));
					add_child(aux2,aux1);
					add_child(out,aux2);
				}
				else
					print_error("expression",curr_index);
			}
			curr_index++;
			add_child(aux,pop_child(out));
			aux1 = parse_term();
			add_child(aux,aux1);
			add_child(tmp,aux);
			add_child(out,tmp);
		}
		else if(is_term(eq_buffer[curr_index])){
			tmp = parse_term();
			add_child(out,tmp);
		}
		else
			break;
	}
	return out;
}

void print_eval_error(char* s){
	printf("Error evaluating %s\n",s);
	exit(0);
}

double eval_constant(enode* node){
	if(cmp_str(node->type,"e\0"))
		return cons_def[0];
	else if(cmp_str(node->type,"pi\0"))
		return cons_def[1];
	else
		print_eval_error("constant");
	return 0;
}

double eval_variable(enode* node){
	if(cmp_str(node->type,"u0\0"))
		return var_def[0];
	else if(cmp_str(node->type,"t\0"))
		return var_def[1];
	else
		print_eval_error("variable");
	return 0;
}

double eval_number(enode* node){
	return atof(node->type);
}

double eval_function(enode* node){
	char* fun = node->type;
	enode* tmp = node->child[0];
	double s = eval_enode(tmp);
	if(cmp_str(fun,"sin"))
		return sin(s);
	else if(cmp_str(fun,"cos"))
		return cos(s);
	else if(cmp_str(fun,"tan"))
		return tan(s);
	else if(cmp_str(fun,"csc"))
		return 1/sin(s);
	else if(cmp_str(fun,"sec"))
		return 1/cos(s);
	else if(cmp_str(fun,"cot"))
		return 1/tan(s);
	else if(cmp_str(fun,"asin"))
		return asin(s);
	else if(cmp_str(fun,"acos"))
		return acos(s);
	else if(cmp_str(fun,"atan"))
		return atan(s);
	else if(cmp_str(fun,"log"))
		return log(s);
	else if(cmp_str(fun,"exp"))
		return exp(s);
	else
		print_eval_error("function");
	return 0;
}

double eval_derivative(enode* node){
	int k = char2number(node->type[1]);
	return deriv_def[k];
}

double eval_term(enode* node){
	return eval_enode(node);
}

double eval_operation(enode* node){
	char* tmp = node->type;
	double s1,s2;
	s1 = eval_enode(node->child[0]);
	s2 = eval_enode(node->child[1]);
	if(cmp_str(tmp,"+\0"))
		return s1+s2;
	else if(cmp_str(tmp,"-\0"))
		return s1-s2;
	else if(cmp_str(tmp,"*\0"))
		return s1*s2;
	else if(cmp_str(tmp,"/\0"))
		return s1/s2;
	else if(cmp_str(tmp,"^\0"))
		return pow(s1,s2);
	return 0;
}

double eval_expression(enode* node){
	return eval_enode(node);
}	

double eval_enode(enode* node){
	if(cmp_str(node->type,"variable"))
		return eval_variable(node->child[0]);
	else if(cmp_str(node->type,"constant"))
		return eval_constant(node->child[0]);
	else if(cmp_str(node->type,"number"))
		return eval_number(node->child[0]);
	else if(cmp_str(node->type,"derivative"))
		return eval_derivative(node->child[0]);
	else if(cmp_str(node->type,"function"))
		return eval_function(node->child[0]);
	else if(cmp_str(node->type,"operation"))
		return eval_operation(node->child[0]);
	else if(cmp_str(node->type,"expression"))
		return eval_expression(node->child[0]);
	else if(cmp_str(node->type,"term"))
		return eval_term(node->child[0]);
}

File: ODESolver/JavaVer/src/com/parse/eParse.java
package com.parse;

import java.util.*;
import com.parse.*;

public class eParse{
	String expr;
	int len;
	int curr_index;
	ArrayList<String> constants;
	ArrayList<String> variables;
	String[] functions = {"sin","cos","tan","csc","sec","cot","log","asin","acos","atan","exp","sinh","cosh","tanh","csch","sech","coth"};
	int degree;

	public eParse(){
		expr = "";
		degree = 0;
	}

	public void setExpr(String s){
		expr = "";
		for(int i = 0; i<s.length();i++){
			if(s.charAt(i) != ' ')
				expr += s.charAt(i);
		}
		len = expr.length();
		curr_index = 0;
	}

	public void setConstants(){
		constants = new ArrayList<String>();
		constants.add("pi");
		constants.add("e");
	}

	public void setVariables(){
		variables = new ArrayList<String>();
		variables.add("x");
	}

	public int getDegree(){
		return degree;
	}

	public boolean isConstant(String s){
		for(String tmp : constants){
			if(tmp.equals(s))
				return true;
		}
		return false;
	}

	public boolean isVariable(String s){
		for(String tmp : variables){
			if(tmp.equals(s))
				return true;
		}
		return false;	
	}

	public boolean isFunction(String s){
		for(String tmp : functions){
			if(tmp.equals(s))
				return true;
		}
		return false;	
	}

	public boolean isDerivative(String s){
		if(s.charAt(0)!='y')
			return false;
		for(int i = 1;i<s.length();i++)
			if(s.charAt(i)!='\'')
				return false;
		return true;
	}

	public boolean isTerm(){
		char c = expr.charAt(curr_index);
		return eUtils.isNumber(c) || eUtils.isLetter(c) || c=='(';	
	}

	public String lookahead(){
		String s = "";
		int index = curr_index;
		char c = expr.charAt(index);
		while(eUtils.isLetter(c) || eUtils.isNumber(c) || c=='\''){
			s += c;
			index++;
			if(index == len)
				break;
			c = expr.charAt(index);
		}
		return s;
	}

	public eNode parseConstant(){
		String s = "";
		eNode node = new eNode("Constant");
		char c = expr.charAt(curr_index);
		while(eUtils.isLetter(c) || eUtils.isNumber(c)){
			s += c;
			curr_index++;
			if(curr_index == len)
				break;
			c = expr.charAt(curr_index);
		}
		node.addChild(new eNode(s));
		return node;
	}

	public eNode parseVariable(){
		String s = "";
		eNode node = new eNode("Variable");
		char c = expr.charAt(curr_index);
		while(eUtils.isLetter(c) || eUtils.isNumber(c)){
			s += c;
			curr_index++;
			if(curr_index == len)
				break;
			c = expr.charAt(curr_index);
		}
		node.addChild(new eNode(s));
		return node;
	}

	public eNode parseNumber(){
		String s = "";
		eNode node = new eNode("Number");
		while(eUtils.isNumber(expr.charAt(curr_index))){
			s += expr.charAt(curr_index);
			curr_index++;
			if(curr_index == len)
				break;
		}
		node.addChild(new eNode(s));
		return node;
	}

	public eNode parseFunction(){
		String s = "";
		eNode node = new eNode("Function");
		eNode tmp,tmp1;
		char c = expr.charAt(curr_index);
		while(eUtils.isLetter(c)){
			s += c;
			curr_index++;
			if(curr_index == len)
				break;
			c = expr.charAt(curr_index);
		}
		if(c != '(')
			System.exit(0);
		curr_index++;
		tmp = parseExpression();
		tmp1 = new eNode(s);
		tmp1.addChild(tmp);
		node.addChild(tmp1);
		c = expr.charAt(curr_index);
		if(c!=')')
			System.exit(0);
		curr_index++;
		return node;
	}

	public eNode parseDerivative(){
		String s = "y";
		int i = 0;
		char c;
		eNode node = new eNode("Derivative");
		curr_index++;
		c = expr.charAt(curr_index);
		while(c=='\''){
			curr_index++;
			i++;
			if(curr_index == len)
				break;
			c = expr.charAt(curr_index);
		}
		s += i;
		if(degree<i)
			degree = i;
		node.addChild(new eNode(s));
		return node;
	}

	public eNode parseSimpleTerm(){
		eNode node = new eNode("Term");
		eNode tmp;
		String str;
		char c = expr.charAt(curr_index);
		if(eUtils.isNumber(c)){
			tmp = parseNumber();
			node.addChild(tmp);
		}
		else if(eUtils.isLetter(c)){
			str = lookahead();
			if(isConstant(str)){
				tmp = parseConstant();
				node.addChild(tmp);
			}
			else if(isVariable(str)){
				tmp = parseVariable();
				node.addChild(tmp);
			}
			else if(isFunction(str)){
				tmp = parseFunction();
				node.addChild(tmp);
			}
			else if(isDerivative(str)){
				tmp = parseDerivative();
				node.addChild(tmp);
			}
		}
		else if(c == '('){
			curr_index++;
			tmp = parseExpression();
			node.addChild(tmp);
			if(expr.charAt(curr_index)!=')')
				System.exit(0);
			curr_index++;
		}
		return node;
	}

	public eNode parseTerm(){
		eNode node = new eNode("Term");
		eNode tmp, tmp1, tmp2;
		String str;
		char c = expr.charAt(curr_index);
		while(!eUtils.isFirstOperation(c) && c!=')'){
			if(eUtils.isNumber(c) || eUtils.isLetter(c) || c == '('){
				tmp = parseSimpleTerm();
				node.addChild(tmp);
			}
			else if(eUtils.isSecondOperation(c)){
				tmp1 = new eNode("Operation");
				tmp2 = new eNode(c+"");
				if(node.getNumChild()==0)
					System.exit(0);
				tmp2.addChild(node.popChild());
				curr_index++;
				tmp = parseSimpleTerm();
				tmp2.addChild(tmp);
				tmp1.addChild(tmp2);
				tmp2 = new eNode("Term");
				tmp2.addChild(tmp1);
				node.addChild(tmp2);
			}
			else{
				System.exit(0);
			}
			if(curr_index == len)
				break;
			c = expr.charAt(curr_index);
		}
		return node;
	}

	public eNode parseExpression(){
		eNode node = new eNode("Expression");
		eNode tmp, tmp1, tmp2,tmp3;
		while(curr_index<len){
			if(eUtils.isFirstOperation(expr.charAt(curr_index))){
				tmp1 = new eNode("Operation");
				tmp2 = new eNode(""+expr.charAt(curr_index));
				if(node.getNumChild() == 0){
					tmp3 = new eNode("Number");
					tmp3.addChild(new eNode("0"));
					tmp = new eNode("Term");
					tmp.addChild(tmp3);
					node.addChild(tmp);
				}
				tmp2.addChild(node.popChild());
				curr_index++;
				tmp3 = parseTerm();
				tmp2.addChild(tmp3);
				tmp1.addChild(tmp2);
				tmp3 = new eNode("Term");
				tmp3.addChild(tmp1);
				node.addChild(tmp3);
			}
			else if(isTerm()){
				tmp = parseTerm();
				node.addChild(tmp);
			}
			else
				break;
		}
		return node;
	}

	public eNode parse(){
		return parseExpression();
	}
}